### PR TITLE
CA-681 support jwt parsing in sam

### PR DESCRIPTION
**Pull request by dvoet**
Ticket: https://broadworkbench.atlassian.net/browse/CA-681

also review https://github.com/broadinstitute/workbench-libs/pull/285

---

**PR checklist**

- [ ] I've followed [the instructions](https://github.com/broadinstitute/sam/blob/develop/CONTRIBUTING.md#api-changes) if I've made any changes to the API, _especially_ if they're breaking changes
- [ ] I've updated the RC_XXX release ticket with any manual steps required to release this change
- [ ] I've updated the [FISMA documentation](https://github.com/broadinstitute/sam/blob/develop/CONTRIBUTING.md#fisma-documentation-changes) if I've made any security-related changes, including auth, encryption, or auditing

### CODE REVIEW

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectives.scala [hunk 1]
```diff
@@ -15,25 +15,31 @@ import org.broadinstitute.dsde.workbench.sam.service.UserService._
...
 trait StandardUserInfoDirectives extends UserInfoDirectives {
   implicit val executionContext: ExecutionContext
 
   def requireUserInfo: Directive1[UserInfo] =
     (
-      headerValueByName(accessTokenHeader) &
-        headerValueByName(googleSubjectIdHeader) &
-        headerValueByName(expiresInHeader) &
-        headerValueByName(emailHeader)
+      optionalHeaderValueByName(accessTokenHeader) &
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectives.scala:27]: the gist of this PR are the changes in this file

File: project/Dependencies.scala [hunk 2]
```diff
@@ -26,6 +26,7 @@ object Dependencies {
   val excludeWorkbenchMetrics = ExclusionRule(organization = "org.broadinstitute.dsde.workbench", name = "workbench-metrics_2.12")
   val excludeWorkbenchGoogle =  ExclusionRule(organization = "org.broadinstitute.dsde.workbench", name = "workbench-google_2.12")
 
+  val jwt: ModuleID = "com.pauldijou" %% "jwt-spray-json" % "4.2.0"
```
Comment (gpolumbo-broad) [inline project/Dependencies.scala:29]: @marctalbott was asking me how I evaluate new modules/plugins/packages from 3rd parties.  As far as I know, there is not a DSP or Broad set of guidelines for selecting 3rd party modules.  I wouldn't want to make this process too formal or unnecessarily restrictive, but maybe at least for our team, we can come up with some guidelines to use?

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectives.scala [hunk 3]
```diff
@@ -74,6 +92,48 @@ object StandardUserInfoDirectives {
...
 
+  def getUserInfoFromJwt(authorizationHeader: String, directoryDAO: DirectoryDAO): IO[UserInfo] = {
+    for {
+      bearerToken <- getBearerTokenFromAuthHeader(authorizationHeader)
+      jwtUserInfo <- parseJwtBearerToken(bearerToken)
+      maybeUser <- directoryDAO.loadUserByIdentityConcentratorId(jwtUserInfo.sub)
+      user <- maybeUser match {
+        case Some(user) => IO.pure(UserInfo(bearerToken, user.id, user.email, jwtUserInfo.exp - Instant.now().getEpochSecond))
+        case None =>
+          // TODO in the next iteration call to IC to find user's google subject id and populate sam_user table
```
Comment (gpolumbo-broad) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectives.scala:103]: https://broadworkbench.atlassian.net/browse/CA-682

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectives.scala [hunk 4]
```diff
@@ -74,6 +92,48 @@ object StandardUserInfoDirectives {
...
+          IO.raiseError(new WorkbenchExceptionWithErrorReport(ErrorReport(StatusCodes.NotFound, s"Identity Concentrator Id ${jwtUserInfo.sub} not found in sam")))
+      }
+    } yield user
+  }
+
+  private def getBearerTokenFromAuthHeader(authorizationHeader: String): IO[OAuth2BearerToken] = {
+    val bearerPrefix = "bearer "
+    // need to be case insensitive when looking for the 'bearer' prefix so can't use stripPrefix
+    if (authorizationHeader.toLowerCase.startsWith(bearerPrefix)) {
```
Comment (gpolumbo-broad) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectives.scala:112]: nit, but seeing since we're parsing this string ourselves, it can't hurt to put a `authorizationHeader.trim()` in there too

File: src/main/resources/org/broadinstitute/dsde/sam/liquibase/changesets/20200212_identity_concentrator_id.xml [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+                   xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog-ext http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-ext.xsd http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.4.xsd">
+
+<changeSet logicalFilePath="dummy" author="dvoet" id="identity_concentrator_id">
+        <addColumn tableName="sam_user">
+            <column name="identity_concentrator_id" type="VARCHAR(30)">
+                <constraints unique="true"/>
+            </column>
+        </addColumn>
+    </changeSet>
+</databaseChangeLog>
```
Comment (yonghaoy) [inline src/main/resources/org/broadinstitute/dsde/sam/liquibase/changesets/20200212_identity_concentrator_id.xml:15]: no new line at end of file

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectives.scala [hunk 6]
```diff
@@ -74,6 +92,48 @@ object StandardUserInfoDirectives {
...
+        case Some(user) => IO.pure(UserInfo(bearerToken, user.id, user.email, jwtUserInfo.exp - Instant.now().getEpochSecond))
+        case None =>
+          // TODO in the next iteration call to IC to find user's google subject id and populate sam_user table
+          IO.raiseError(new WorkbenchExceptionWithErrorReport(ErrorReport(StatusCodes.NotFound, s"Identity Concentrator Id ${jwtUserInfo.sub} not found in sam")))
+      }
+    } yield user
+  }
+
+  private def getBearerTokenFromAuthHeader(authorizationHeader: String): IO[OAuth2BearerToken] = {
+    val bearerPrefix = "bearer "
```
Comment (yonghaoy) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectives.scala:110]: Is that possible that we have extra space around Bear? Like "Bearer  "?

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectives.scala [hunk 7]
```diff
@@ -74,6 +92,48 @@ object StandardUserInfoDirectives {
...
+    // need to be case insensitive when looking for the 'bearer' prefix so can't use stripPrefix
+    if (authorizationHeader.toLowerCase.startsWith(bearerPrefix)) {
+      IO.pure(OAuth2BearerToken(authorizationHeader.drop(bearerPrefix.size)))
```
Comment (yonghaoy) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectives.scala:113]: Just to confirm, authorizationHeader is not changed to lower case by the line above?

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/directory/PostgresDirectoryDAO.scala [hunk 8]
```diff
@@ -470,7 +470,17 @@ class PostgresDirectoryDAO(protected val dbRef: DbReference,
 
       val loadUserQuery = samsql"select ${userTable.resultAll} from ${UserTable as userTable} where ${userTable.id} = ${userId}"
       loadUserQuery.map(UserTable(userTable))
-        .single().apply().map(unmarshalUserRecord)
+        .single().apply().map(UserTable.unmarshalUserRecord)
+    }
+  }
+
+  override def loadUserByIdentityConcentratorId(userId: IdentityConcentratorId): IO[Option[WorkbenchUser]] = {
```
Comment (yonghaoy) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/directory/PostgresDirectoryDAO.scala:477]: maybe not related with this PR, but something we need to consider sometime in the future is : 
Since IC support link two google accounts, and in this case, it technically possible that two SAM user records would have the same IC id. And we would get multiple entities here. 
Do we need to consider that? Like catch exception during .single() and log something for now?

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/openam/PostgresAccessPolicyDAO.scala [hunk 9]
```diff
@@ -756,14 +756,11 @@ class PostgresAccessPolicyDAO(protected val dbRef: DbReference,
 
     runInTransaction { implicit session =>
       val query = samsql"""with recursive ${recursiveMembersQuery(policyId, subGroupMemberTable)}
-        select ${sg.result.memberUserId}, ${u.result.googleSubjectId}, ${u.result.email}
+        select ${u.resultAll}
```
Comment (yonghaoy) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/openam/PostgresAccessPolicyDAO.scala:759]: What the purpose of this change? Is this related with jwt? For me it seems for dedup query results only...

File: src/test/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectivesSpec.scala [hunk 10]
```diff
@@ -42,7 +42,7 @@ class StandardUserInfoDirectivesSpec extends FlatSpec with PropertyBasedTesting
       val directoryDAO = new MockDirectoryDAO()
       val googleSubjectId = GoogleSubjectId(genRandom(System.currentTimeMillis()))
       val uid = genWorkbenchUserId(System.currentTimeMillis())
-      directoryDAO.createUser(WorkbenchUser(uid, Some(googleSubjectId), email)).unsafeRunSync()
+      directoryDAO.createUser(WorkbenchUser(uid, Some(googleSubjectId), email, None)).unsafeRunSync()
```
Comment (yonghaoy) [inline src/test/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectivesSpec.scala:45]: Do we need to create user here with some icId instead of none?

File: src/test/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectivesSpec.scala [hunk 11]
```diff
@@ -53,7 +53,7 @@ class StandardUserInfoDirectivesSpec extends FlatSpec with PropertyBasedTesting
       (serviceSubjectId: ServiceAccountSubjectId, googleSubjectId: GoogleSubjectId, token: OAuth2BearerToken, email: WorkbenchEmail) =>
       val directoryDAO = new MockDirectoryDAO()
       val uid = genWorkbenchUserId(System.currentTimeMillis())
-      directoryDAO.createUser(WorkbenchUser(uid, Some(googleSubjectId), email)).unsafeRunSync()
+      directoryDAO.createUser(WorkbenchUser(uid, Some(googleSubjectId), email, None)).unsafeRunSync()
```
Comment (yonghaoy) [inline src/test/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectivesSpec.scala:56]: ditto

File: src/test/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectivesSpec.scala [hunk 12]
```diff
@@ -105,6 +105,53 @@ class StandardUserInfoDirectivesSpec extends FlatSpec with PropertyBasedTesting
...
...iOlsiOTAzY2ZhZWItNTdkOS00ZWY2LTU2NTktMDQzNzc3OTRlZDY1Il0sImNsaWVudF9pZCI6IjkwM2NmYWViLTU3ZDktNGVmNi01NjU5LTA0Mzc3Nzk0ZWQ2NSIsImV4cCI6MTU4MTYwNTIzMCwiZXh0Ijp7ImlkZW50aXRpZXMiOlsiZHZvZXR0ZXN0QGdtYWlsLmNvbSJdfSwiaWF0IjoxNTgxNjAxNjI5LCJpc3MiOiJodHRwczovL2ljLWRvdC1kb2dlaWN0ZXN0LmFwcHNwb3QuY29tLyIsImp0aSI6Ijc0ZGFkNTFkLTEwZTAtNDk4OC1iOGRlLWI2YTkzYmExODcxMCIsIm5iZiI6MTU4MTYwMTYyOSwic2NwIjpbIm9wZW5pZCIsIm9mZmxpbmUiLCJwcm9maWxlIiwiaWRlbnRpdGllcyIsImdhNGdoX3Bhc3Nwb3J0X3YxIl0sInN1YiI6ImljX2ZiOGMyNGYxMWNkYjQ5MGU4ODE1NzQifQ.B7KUVxofXA_5PBgIN4qEIBl3vhXntgoGKNJ0AE-FzC2X_vLLBSeZ4FpGIjHjFr75XHG0-kPU1wwiz7_wmwsJpwMhybvRyn5f2PjFfzf6HrD3cEkJiz0PnIxkH4Eb3xksOJMhAfZh1ZQlbKXdcc54oDmM3JTyu-YqIQT1J1kBmAqWM3Wy_YX2KtSZHJTQHVxDhFdq9EQosraNJT7wH9lFPPZmNq5YXQVOw7V4lB9UBOFC4sSwnS332xPN_tmzxcobmzEVHpA3UwKI1QfxXQRLxpllavlizN80hJlQgXGIk-GYgFe92JpNbhx2H0ipiUj16e7uKpmwtI3myoXPkQ8wOeVB2ofuLP-FZxojLY4fuUEUA7hg0Gmu6dvV9sz5bVHmxlaIQQl7Pjvz5FlJ7fWWOOTvOLjwMNM546b9Oe_tdovv2cqyQI4f6p6Q0Bk5O7v4-BCd8RWHgWTaA_-GQ0ixn26YUkFUyCilhR4TFJE_0t1Dsyb-wlwc1Op0oTjfilpRlCaOT9AhsMnLWqvtKFhMhIis2_vDCwNSDfNwNXDe-rTCMPqxR42jCfpBOQ8SPSs8YHyLYCGXPTDnsm0zwYg2imM-palKBeJM6RGkQOzBkCA3T4IYuDnxZJekpxGbSluPeruPwc02EjB7Q2qWHw3jnPPSTslwjsw1IieCcxdFmUc"
```
Comment (yonghaoy) [inline src/test/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectivesSpec.scala]: Optional: This is confused because we don't know how those jwt are created and if they actually decodable. Is that easy to generate jwt from json or somewhere, and encode it here? 
Feel free to push it back if that is a none trial change..

File: src/test/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectivesSpec.scala [hunk 13]
```diff
@@ -105,6 +105,53 @@ class StandardUserInfoDirectivesSpec extends FlatSpec with PropertyBasedTesting
...
+
+  "getUserInfoFromJwt" should "get user info for valid jwt" in {
+    val directoryDAO = new MockDirectoryDAO()
+    val googleSubjectId = GoogleSubjectId(genRandom(System.currentTimeMillis()))
+    val uid = genWorkbenchUserId(System.currentTimeMillis())
+    val someIdentityConcentratorId = Some(IdentityConcentratorId("ic_fb8c24f11cdb490e881574"))
```
Comment (yonghaoy) [inline src/test/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectivesSpec.scala:115]: is the ic id some random value you created?

File: src/test/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectivesSpec.scala [hunk 14]
```diff
@@ -105,6 +105,53 @@ class StandardUserInfoDirectivesSpec extends FlatSpec with PropertyBasedTesting
...
+
+  it should "404 for valid jwt but user does not exist" in {
+    val directoryDAO = new MockDirectoryDAO()
+    val t = intercept[WorkbenchExceptionWithErrorReport] {
+      getUserInfoFromJwt(s"Bearer $validJwt", directoryDAO).unsafeRunSync()
+    }
+    t.errorReport.statusCode shouldBe Some(StatusCodes.NotFound)
+  }
+
+  it should "400 with no exp" in {
```
Comment (yonghaoy) [inline src/test/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectivesSpec.scala]: I thought we are throwing 401 code - Unauthorized?

File: src/test/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectivesSpec.scala [hunk 15]
```diff
@@ -105,6 +105,53 @@ class StandardUserInfoDirectivesSpec extends FlatSpec with PropertyBasedTesting
...
+    val directoryDAO = new MockDirectoryDAO()
+    val t = intercept[WorkbenchExceptionWithErrorReport] {
+      getUserInfoFromJwt(s"Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTE2MjM5MDIyfQ.SflKxwRJSMeKKF2QT4fwpMeJf36POk6yJV_adQssw5c", directoryDAO).unsafeRunSync()
+    }
+    t.errorReport.statusCode shouldBe Some(StatusCodes.Unauthorized)
+  }
+
+  it should "400 with no sub" in {
```
Comment (yonghaoy) [inline src/test/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectivesSpec.scala]: ditto and same for the 400 below

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectives.scala [hunk 16]
```diff
@@ -15,25 +15,31 @@ import org.broadinstitute.dsde.workbench.sam.service.UserService._
...
     ) tflatMap {
-      case (token, googleSubjectId, expiresIn, email) =>
-        onSuccess(
-          Try(expiresIn.toLong).fold[Future[UserInfo]](
-            t =>
-              Future.failed(
-                new WorkbenchExceptionWithErrorReport(ErrorReport(StatusCodes.BadRequest, s"expiresIn $expiresIn can't be converted to Long because  of $t"))),
-            l => getUserInfo(OAuth2BearerToken(token), GoogleSubjectId(googleSubjectId), WorkbenchEmail(email), l, directoryDAO).unsafeToFuture()
-          ))
+      case (Some(token), Some(googleSubjectId), Some(expiresIn), Some(email), _) =>
```
Comment (yonghaoy) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectives.scala:33]: Is that possible we are receiving OidcHeaders comes with authorizationHeader?

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectives.scala [hunk 17]
```diff
@@ -15,25 +15,31 @@ import org.broadinstitute.dsde.workbench.sam.service.UserService._
...
+        // request coming through Apache proxy
+        onSuccess {
+          getUserInfoFromOidcHeaders(directoryDAO, googleSubjectId, expiresIn, email, OAuth2BearerToken(token)).unsafeToFuture()
+        }
+      case (_, _, _, _, authHeader) =>
+        // JWT issued by Identity Concentrator - already validated by proxy
+        onSuccess {
+          getUserInfoFromJwt(authHeader, directoryDAO).unsafeToFuture()
+        }
```
Comment (yonghaoy) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectives.scala:42]: Do we need case _ => "other" and error out something here?

**Review by yonghaoy — COMMENTED** [review]
This PR looks great! Most of my comments just for my learning purpose. Feel free to push anything back! 
Thanks

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectives.scala (hunk 6 above)
Comment (dvoet) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectives.scala:110]: I am not sure what you mean

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectives.scala (hunk 7 above)
Comment (dvoet) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectives.scala:113]: correct, lowercase is just for comparison. The token itself is case sensitive.

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/directory/PostgresDirectoryDAO.scala (hunk 8 above)
Comment (dvoet) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/directory/PostgresDirectoryDAO.scala:477]: The first phase of releasing IC will support only google and no account linking. Once we are fully off of opaque google tokens we can figure out how we want to support account linking.

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/openam/PostgresAccessPolicyDAO.scala (hunk 9 above)
Comment (dvoet) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/openam/PostgresAccessPolicyDAO.scala:759]: just some code cleanup

File: src/test/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectivesSpec.scala (hunk 10 above)
Comment (dvoet) [reply src/test/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectivesSpec.scala:45]: this test does not care

File: src/test/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectivesSpec.scala (hunk 12 above)
Comment (dvoet) [reply src/test/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectivesSpec.scala]: I did this specifically because it was generated by IC... I will add a comment to that effect.

File: src/test/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectivesSpec.scala (hunk 14 above)
Comment (dvoet) [reply src/test/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectivesSpec.scala]: indeed, will fix

File: src/test/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectivesSpec.scala (hunk 13 above)
Comment (dvoet) [reply src/test/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectivesSpec.scala:115]: came from a token I got from IC

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectives.scala (hunk 16 above)
Comment (dvoet) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectives.scala:33]: yes and it definitely will. The `_` means that we don't care what it is

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectives.scala (hunk 17 above)
Comment (dvoet) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectives.scala:42]: no, the 2 cases are exhaustive (and the compiler would tell us otherwise)

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectives.scala (hunk 6 above)
Comment (yonghaoy) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectives.scala:110]: Sorry, I mean extra spaces instead of just one, like "Bear     "?

**Review by yonghaoy — APPROVED** [review]

**Review by gpolumbo-broad — APPROVED** [review]

**Review by gpolumbo-broad — CHANGES_REQUESTED** [review]
@dvoet - it looked like there were a few comments from yonghao that you said you would address.  In particular in the tests where we were going to return 401s instead of 400s.  Barring those corrections, I'm 👍

**Review by gpolumbo-broad — APPROVED** [review]

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectives.scala (hunk 4 above)
Comment (dvoet) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectives.scala:112]: I am not sure about that... the spec is pretty specific

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectives.scala (hunk 6 above)
Comment (dvoet) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectives.scala:110]: I am pretty certain the code as written follows the spec for how bearer tokens should be specified. There can be only 1 space after `bearer`

**Comment by dvoet** [comment]
jenkins retest - prior run had failure downloading some sbt dependencies
